### PR TITLE
[5/n] frontend for url code references

### PR DESCRIPTION
**Pull request by benpankow**
## Summary

Adds GQL layer, frontend for source control links. These render with the icon for the corresponding source control service and hyperlink to the targeted URL.

Screenshot 2024-05-07 at 5 02 05 PM

Screenshot 2024-05-07 at 5 02 30 PM

Screenshot 2024-05-07 at 5 06 40 PM

## Test Plan

Tested locally.

### CODE REVIEW

**Comment by benpankow** [comment]
> [!WARNING]
> <b>This pull request is not mergeable via GitHub because a downstack PR is open. Once all requirements are satisfied, merge this PR as a stack <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21713?utm_source=stack-comment-downstack-mergeability-warning" >on Graphite</a>.</b>
> <a href="https://graphite.dev/docs/merge-pull-requests">Learn more</a>

* **#21713** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21713?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* **#21675** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21675?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21468** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21468?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21467** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21467?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21465** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21465?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#21576** <a href="https://app.graphite.dev/github/pr/dagster-io/dagster/21576?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* `master`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @benpankow and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: js_modules/dagster-ui/packages/ui-core/src/metadata/__stories__/MetadataEntry.stories.tsx [hunk 1]
```diff
@@ -270,6 +270,11 @@ function buildMockMetadataEntry(type: MetadataEntryFragment['__typename']): Meta
             lineNumber: 12,
             label: 'my_code_reference',
           },
+          {
+            __typename: 'UrlCodeReference',
```
Comment (salazarm) [inline js_modules/dagster-ui/packages/ui-core/src/metadata/__stories__/MetadataEntry.stories.tsx:274]: buildUrlCodeReference()

**Review by salazarm — APPROVED** [review]

File: js_modules/dagster-ui/packages/ui-core/src/code-links/CodeLink.tsx [hunk 2]
```diff
@@ -34,28 +84,38 @@ export const CodeLink = ({codeLinkData}: {codeLinkData: CodeReferencesMetadataEn
...
+                >
+                  <MenuItem
+                    text={getCodeReferenceEntryLabel(source)}
+                    onClick={() => {
+                      const codeLink = getCodeReferenceLink(codeLinkProtocol, source);
+                      window.open(codeLink, '_blank');
+                    }}
+                    icon={<Icon name={getCodeReferenceIcon(source)} />}
+                    style={{maxWidth: 300}}
+                  />
```
Comment (salazarm) [inline js_modules/dagster-ui/packages/ui-core/src/code-links/CodeLink.tsx:101]: This could also be a MenuLink

**Comment by benpankow** [comment]
### Merge activity

* **May 15, 2:09 PM PDT**: @benpankow started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/dagster-io/dagster/21713).

**Pull request merged**